docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,339 @@
+
+```go
+type Configurator interface {
+   MsgServer() grpc.Server
+   QueryServer() grpc.Server
+   InternalServer() grpc.Server
+}
+```
+
+Services registered against `InternalServer` will be callable from other modules but not by external clients.
+
+### Authorization
+
+By default, the inter-module router requires that messages are sent by the first signer returned by `GetSigners`. The
+inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).
+This middleware will allow accounts to otherwise specific module accounts to perform actions on their behalf.
+Authorization middleware should take into account the need to grant certain modules effectively "admin" privileges to
+other modules. This will be addressed in separate ADRs or updates to this ADR.
+
+### Future Work
+
+Other future improvements may include:
+* custom code generation that:
+  * simplifies interfaces (ex. generates code with `sdk.Context` instead of `context.Context`)
+  * optimizes inter-module calls - for instance caching resolved methods after first invocation
+* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
+* code generation which makes inter-module communication more performant
+* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
+* inter-module hooks and plugins
+
+## Alternatives
+
+The `x/capability` module does provide a proper object-capability implementation that can be used by any module in the
+SDK and could even be used for inter-module Ocaps as described in [\#5931](https://github.com/cosmos/cosmos-sdk/issues/5931).
+
+The advantages of the approach described in this ADR are mostly around how it integrates with other parts of the SDK,
+specifically:
+
+* protobuf so that:
+  * code generation of interfaces can be leveraged for a better dev UX
+  * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)
+* sub-module accounts as per ADR 028
+* the general `Msg` passing paradigm and the way signers are specified by `GetSigners`
+
+Also, this is a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the
+`x/capability` approach in #5931 would need to be applied method by method.
+
+## Consequences
+
+### Backwards Compatibility
+
+This ADR is intended to provide a pathway to a scenario where there is greater long term compatibility between modules.
+In the short-term, this will likely result in breaking certain `Keeper` interfaces which are too permissive and/or
+replacing `Keeper` interfaces altogether.
+
+### Positive
+
+- an alternative to keepers which can more easily lead to stable inter-module interfaces
+- proper inter-module Ocaps
+
+### Negative
+
+- modules which adopt this will need significant refactoring
```

Comment:
I'd prefer an alternative that doesn't cause this amount of refactoring. The dev UX gain from this ADR is probably not enough to make it worth for SDK devs to implement this solution. Stargate has a massive amount of changes, and I honestly think we should reconsider the amount of breaking changes that new ADRs generate.

Reply:
Commented on the alternative a few places, namely _converting_ is a different discussion from _enabling_.

Reply:
I agree.... there is a huge exhaustion of following breaking changes *every sdk release* for 2 years now.

It is not just stargate, but even 0.34 -> 0.36 or 0.37 -> 0.38 upgrades were quite painful. With the whole layout of the modules restructured (and re-restructured) as the core devs had a new, inspiring design. The change-fatigue and desire for stability in the ecosystem should be taking into consideration.

There are also many projects on mainnet with many custom modules and only a few of them are rushing to stargate. The more backwards-incompatible changes, the more we risk splintering the ecosystem.